File: ssm-sdk-client/src/main/java/ssm/client/repository/RepositoryFactory.java
package ssm.client.repository;

import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import retrofit2.Retrofit;
import retrofit2.converter.jackson.JacksonConverterFactory;

public class RepositoryFactory {

    private static OkHttpClient.Builder httpClient;

    private final Retrofit retrofit;

    public RepositoryFactory(String baseUrl) {
        this.retrofit = buildRetrofit(baseUrl);
    }

    private Retrofit buildRetrofit(String baseUrl) {
        initializeHttpBuilder();
        OkHttpClient client = httpClient.build();

        return new Retrofit.Builder()
                .baseUrl(baseUrl)
                .addConverterFactory(JacksonConverterFactory.create())
                .client(client)
                .build();
    }

    private void initializeHttpBuilder() {
        if (httpClient != null) {
            return;
        }

        httpClient = new OkHttpClient.Builder()
            .addInterceptor(chain -> {
                Request original = chain.request();
                Request.Builder requestBuilder = original.newBuilder()
                    .header("Content-Type", "application/json").method(original.method(), original.body());
                return chain.proceed(requestBuilder.build());
            })
            .connectTimeout(20, TimeUnit.SECONDS)
            .readTimeout(20, TimeUnit.SECONDS)
            .writeTimeout(20, TimeUnit.SECONDS);
    }

    public CoopRepository buildCoopRepository() {
        return retrofit.create(CoopRepository.class);
    }
}

File: ssm-sdk/ssm-sdk-core/src/main/kotlin/ssm/chaincode/client/repository/RepositoryFactory.kt
package ssm.chaincode.client.repository

import java.util.concurrent.TimeUnit
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.jackson.JacksonConverterFactory

class RepositoryFactory(baseUrl: String, bearerToken: String?) {

	private val retrofit: Retrofit

	companion object {
		const val TIMEOUT: Long = 20
		private var httpClient: OkHttpClient.Builder? = null
	}

	init {
		retrofit = buildRetrofit(baseUrl, bearerToken)
	}

	private fun buildRetrofit(baseUrl: String, bearerToken: String?): Retrofit {
		val client = buildHttpClient(bearerToken)
		return Retrofit.Builder()
			.baseUrl(baseUrl)
			.addConverterFactory(JacksonConverterFactory.create())
			.client(client)
			.build()
	}

	private fun buildHttpClient(bearerToken: String?): OkHttpClient {
		initializeHttpBuilder()
		httpClient!!.interceptors().clear()
		return httpClient!!.addInterceptor { chain: Interceptor.Chain ->
			val original = chain.request()
			var requestBuilder = original.newBuilder()
				.header("Content-Type", "application/json").method(original.method(), original.body())
			if (bearerToken != null && !bearerToken.isEmpty()) {
				requestBuilder = requestBuilder.header("Authorization", "Bearer $bearerToken")
			}
			chain.proceed(requestBuilder.build())
		}.build()
	}

	private fun initializeHttpBuilder() {
		if (httpClient != null) {
			return
		}
		httpClient = OkHttpClient.Builder()
			.connectTimeout(TIMEOUT, TimeUnit.SECONDS)
			.readTimeout(TIMEOUT, TimeUnit.SECONDS)
			.writeTimeout(TIMEOUT, TimeUnit.SECONDS)
	}

	fun buildCoopRepository(): CoopRepository {
		return retrofit.create(CoopRepository::class.java)
	}
}
